src/image-tvg.c

```diff
@@ -0,0 +1,1379 @@
+    TVG_RANGE_ENHANCED,
+};
+
+/* The color encoding used in a TinyVG file.
+ * This enum describes how the data in the color table
+ * section of the format looks like.
+ */
+enum {
+    TVG_COLOR_U8888 = 0,
+    TVG_COLOR_U565,
+    TVG_COLOR_F32,
+    TVG_COLOR_CUSTOM,
+};
+
+/* A TinyVG scale value. Defines the scale for all units inside a graphic.
+ * The scale is defined by the number of decimal bits in a `i32`, thus scaling
+ * can be trivially implemented by shifting the integers right by the scale
+ * bits.
+ */
+enum {
+    TVG_SCALE_1_1 = 0,
+    TVG_SCALE_1_2,
+    TVG_SCALE_1_4,
+    TVG_SCALE_1_8,
+    TVG_SCALE_1_16,
+    TVG_SCALE_1_32,
+    TVG_SCALE_1_64,
+    TVG_SCALE_1_128,
+    TVG_SCALE_1_256,
+    TVG_SCALE_1_512,
+    TVG_SCALE_1_1024,
+    TVG_SCALE_1_2048,
+    TVG_SCALE_1_4096,
+    TVG_SCALE_1_8192,
+    TVG_SCALE_1_16384,
+    TVG_SCALE_1_32768,
+};
+
+/* path commands */
+enum {
+    TVG_PATH_LINE = 0,
+    TVG_PATH_HLINE,
+    TVG_PATH_VLINE,
+    TVG_PATH_CUBIC,
+    TVG_PATH_ARC_CIRCLE,
+    TVG_PATH_ARC_ELLIPSE,
+    TVG_PATH_CLOSE,
+    TVG_PATH_QUAD
+};
+
+enum {
+    TVG_SUCCESS = 0,
+    TVG_E_INVALID_ARG,
+    TVG_E_INVALID_STATE,
+    TVG_E_INVALID_FORMAT,
+    TVG_E_IO_ERROR,
+    TVG_E_OUT_OF_MEMORY,
+    TVG_E_NOT_SUPPORTED
+};
+
+/* clamp a value to a range */
+#define TVG_CLAMP(x, mn, mx) (x > mx ? mx : (x < mn ? mn : x))
+/* get the red channel of an RGB565 color */
+#define TVG_RGB16_R(x) (x & 0x1F)
+/* get the green channel of an RGB565 color */
+#define TVG_RGB16_G(x) ((x >> 5) & 0x3F)
+/* get the blue channel of an RGB565 color */
+#define TVG_RGB16_B(x) ((x >> 11) & 0x1F)
+/* get the index of the command
+ * essentially the command id
+ */
+#define TVG_CMD_INDEX(x) (x & 0x3F)
+/* get the style kind flags in the command */
+#define TVG_CMD_STYLE_KIND(x) ((x >> 6) & 0x3)
+/* get the packed size out of the size
+ * and style kind packed value
+ */
+#define TVG_SIZE_AND_STYLE_SIZE(x) ((x & 0x3F) + 1)
+/* get the style kind out of the size
+ * and style kind packed value
+ */
+#define TVG_SIZE_AND_STYLE_STYLE_KIND(x) ((x >> 6) & 0x3)
+/* get the scale from the header */
+#define TVG_HEADER_DATA_SCALE(x) (x & 0x0F)
+/* get the color encoding from the header */
+#define TVG_HEADER_DATA_COLOR_ENC(x) ((x >> 4) & 0x03)
+/* get the color range from the header */
+#define TVG_HEADER_DATA_RANGE(x) ((x >> 6) & 0x03)
+/* get the path command index/id */
+#define TVG_PATH_CMD_INDEX(x) (x & 0x7)
+/* flag indicating the path has line/stroke to it */
+#define TVG_PATH_CMD_HAS_LINE(x) ((x >> 4) & 0x1)
+/* flag indicating the arc is a large arc */
+#define TVG_ARC_LARGE(x) (x & 0x1)
+/* flag indicating the sweep direction 0=left, 1=right */
+#define TVG_ARC_SWEEP(x) ((x >> 1) & 1)
+
+/* F32 pixel format */
+typedef struct {
+    float r, g, b, a;
+} tvg_f32_pixel_t;
+
+/* rgba32 color struct */
+typedef struct {
+    uint8_t r;
+    uint8_t g;
+    uint8_t b;
+    uint8_t a;
+} tvg_rgba32_t;
```

Comment:
`typedef uint32_t twin_argb32_t;` 
Can we reuse the previous implementation? I'm uncertain if it's better to stick with the previous type by converting from `float` to `uint8_t` when the pixel format is `float`.